rust_src/src/editfns.rs

```diff
@@ -945,4 +942,155 @@ pub fn buffer_string() -> LispObject {
+// the value of point is used instead.  If BEG or END is null,
+// means don't store the beginning or end of the field.
+//
+// BEG_LIMIT and END_LIMIT serve to limit the ranged of the returned
+// results; they do not effect boundary behavior.
+//
+// If MERGE_AT_BOUNDARY is non-nil, then if POS is at the very first
+// position of a field, then the beginning of the previous field is
+// returned instead of the beginning of POS's field (since the end of a
+// field is actually also the beginning of the next input field, this
+// behavior is sometimes useful).  Additionally in the MERGE_AT_BOUNDARY
+// non-nil case, if two fields are separated by a field with the special
+// value `boundary', and POS lies within it, then the two separated
+// fields are considered to be adjacent, and POS between them, when
+// finding the beginning and ending of the "merged" field.
+//
+// Either BEG or END may be 0, in which case the corresponding value
+// is not stored.
+
+#[no_mangle]
+pub extern "C" fn find_field(
+    pos: LispObject,
+    merge_at_boundary: LispObject,
+    beg_limit: LispObject,
+    beg: *mut ptrdiff_t,
+    end_limit: LispObject,
+    end: *mut ptrdiff_t,
+) {
+    rust_find_field(
+        pos.into(),
+        merge_at_boundary.into(),
+        beg_limit,
+        beg,
+        end_limit,
+        end,
+    );
+}
+
+pub fn rust_find_field(
+    pos: Option<LispNumber>,
+    merge_at_boundary: bool,
+    beg_limit: LispObject,
+    beg: *mut ptrdiff_t,
+    end_limit: LispObject,
+    end: *mut ptrdiff_t,
+) {
+    let current_buffer = ThreadState::current_buffer();
+    let pos = pos.map_or_else(|| current_buffer.pt, |p| p.to_fixnum() as isize);
+
+    // Fields right before and after the point.
+    let after_field =
+        unsafe { get_char_property_and_overlay(pos.into(), Qfield, Qnil, ptr::null_mut()) };
+    let before_field = if pos > current_buffer.begv {
+        unsafe { get_char_property_and_overlay((pos - 1).into(), Qfield, Qnil, ptr::null_mut()) }
+    } else {
+        // Using nil here would be a more obvious choice, but it would
+        // fail when the buffer starts with a non-sticky field.
+        after_field
+    };
+
+    // True if POS counts as the start of a field.
+    let mut at_field_start = false;
+    // True if POS counts as the end of a field.
+    let mut at_field_end = false;
+
+    // See if we need to handle the case where MERGE_AT_BOUNDARY is nil
+    // and POS is at beginning of a field, which can also be interpreted
+    // as the end of the previous field.  Note that the case where if
+    // MERGE_AT_BOUNDARY is non-nil (see function comment) is actually the
+    // more natural one; then we avoid treating the beginning of a field
+    // specially.
+    if !merge_at_boundary {
+        let field = unsafe { Fget_pos_property(pos.into(), Qfield, Qnil) };
+        if !field.eq(after_field) {
+            at_field_end = true;
+        }
+        if !field.eq(before_field) {
```

Comment:
You can use `ne` instead of `eq`.

Reply:
yeah, I am not sold on `ne` or `is_not_nil`.

Reply:
Ok. I think it's not important.